File: zig/1722.zig
const std = @import("std");

const UnionFind = struct {
    parent: []usize,
    rank: []usize,

    fn new(allocator: std.mem.Allocator, len: usize) !UnionFind {
        var parent = try allocator.alloc(usize, len);
        var rank = try allocator.alloc(usize, len);

        for (0..len) |i| {
            parent[i] = i;
            rank[i] = 1;
        }

        return UnionFind{
            .parent = parent,
            .rank = rank,
        };
    }

    fn deinit(self: *UnionFind, allocator: std.mem.Allocator) void {
        allocator.free(self.parent);
        allocator.free(self.rank);
    }

    fn find(self: *UnionFind, x: usize) usize {
        if (self.parent[x] != x) {
            self.parent[x] = self.find(self.parent[x]);
        }

        return self.parent[x];
    }

    fn unite(self: *UnionFind, a: usize, b: usize) void {
        const rootA = self.find(a);
        const rootB = self.find(b);

        if (rootA != rootB) {
            if (self.rank[rootA] > self.rank[rootB]) {
                self.parent[rootB] = rootA;
            } else if (self.rank[rootA] < self.rank[rootB]) {
                self.parent[rootA] = rootB;
            } else {
                self.parent[rootB] = rootA;
                self.rank[rootA] += 1;
            }
        }
    }
};

fn solution(allocator: std.mem.Allocator, source: []const usize, target: []const usize, allowedSwaps: []const [2]usize) !usize {
    const n = source.len;
    var uf = try UnionFind.new(allocator, n);

    for (allowedSwaps) |swap| {
        uf.unite(swap[0], swap[1]);
    }

    var mp: std.AutoHashMap(usize, std.AutoHashMap(usize, usize)) = .init(allocator);
    defer {
        uf.deinit(allocator);

        var it = mp.iterator();
        while (it.next()) |entry| {
            entry.value_ptr.*.deinit();
        }

        mp.deinit();
    }

    for (0..n) |i| {
        const root = uf.find(i);

        const gop = try mp.getOrPut(root);
        if (!gop.found_existing) {
            gop.value_ptr.* = std.AutoHashMap(usize, usize).init(allocator);
        }

        const freq_map = gop.value_ptr;
        const entry = try freq_map.getOrPut(source[i]);
        if (!entry.found_existing) {
            entry.value_ptr.* = 0;
        }

        entry.value_ptr.* += 1;
    }

    var hamming_dist: usize = 0;
    for (0..n) |i| {
        const root = uf.find(i);
        var freq_map = mp.getPtr(root).?;

        if (freq_map.getPtr(target[i])) |count_ptr| {
            if (count_ptr.* > 0) {
                count_ptr.* -= 1;
            } else {
                hamming_dist += 1;
            }
        } else {
            hamming_dist += 1;
        }
    }

    return hamming_dist;
}

const expect = std.testing.expectEqual;

test "test case #1" {
    const allocator = std.testing.allocator;

    const source = [_]usize{ 1, 2, 3, 4 };
    const target = [_]usize{ 2, 1, 4, 5 };
    const allowedSwaps = [_][2]usize{
        .{ 0, 1 },
        .{ 2, 3 },
    };

    const expected = 1;
    const actual = try solution(allocator, &source, &target, &allowedSwaps);

    try expect(expected, actual);
}

test "test case #2" {
    const allocator = std.testing.allocator;

    const source = [_]usize{ 1, 2, 3, 4 };
    const target = [_]usize{ 1, 3, 2, 4 };
    const allowedSwaps = [_][2]usize{};

    const expected = 2;
    const actual = try solution(allocator, &source, &target, &allowedSwaps);

    try expect(expected, actual);
}

test "test case #3" {
    const allocator = std.testing.allocator;

    const source = [_]usize{ 5, 1, 2, 4, 3 };
    const target = [_]usize{ 1, 5, 4, 2, 3 };
    const allowedSwaps = [_][2]usize{
        .{ 0, 4 },
        .{ 4, 2 },
        .{ 1, 3 },
        .{ 1, 4 },
    };

    const expected = 0;
    const actual = try solution(allocator, &source, &target, &allowedSwaps);

    try expect(expected, actual);
}

File: go/1722.go
package main

type UnionFind struct {
	parent []int
	rank   []int
}

func New(len int) UnionFind {
	parent := make([]int, len)
	rank := make([]int, len)
	for i := 0; i < len; i++ {
		parent[i] = i
		rank[i] = 0
	}

	return UnionFind{
		parent: parent,
		rank:   rank,
	}
}

func (u *UnionFind) find(x int) int {
	if u.parent[x] != x {
		u.parent[x] = u.find(u.parent[x])
	}

	return u.parent[x]
}

func (u *UnionFind) union(a, b int) {
	parentA := u.find(a)
	parentB := u.find(b)

	if parentA != parentB {
		if u.rank[parentA] > u.rank[parentB] {
			u.parent[parentB] = parentA
		} else if u.rank[parentA] < u.rank[parentB] {
			u.parent[parentA] = parentB
		} else {
			u.parent[parentB] = parentA
			u.rank[parentA]++
		}
	}
}

func minimumHammingDistance(source []int, target []int, allowedSwaps [][]int) int {
	n := len(source)
	uf := New(n)

	for _, swap := range allowedSwaps {
		uf.union(swap[0], swap[1])
	}

	mp := map[int]map[int]int{}
	for i := 0; i < n; i++ {
		root := uf.find(i)

		if mp[root] == nil {
			mp[root] = make(map[int]int)
		}

		mp[root][source[i]]++
	}

	result := 0
	for i := 0; i < n; i++ {
		root := uf.find(i)
		freq := mp[root]

		if count, ok := freq[target[i]]; ok && count > 0 {
			freq[target[i]]--
		} else {
			result++
		}
	}

	return result
}
